Context free algebras

GrafBlutwurst opened the pull request:
I think we're hitting the limit of scalac I only got the scala compiler to shut up by expicitly disabling erasure warning though. regular patterns are not possible because the compiler cannot infer the types correctly and adding the types by hand triggers erasure warnings.

GrafBlutwurst committed:
added initial prototype for deriving context free algebras from typeclasses

GrafBlutwurst committed:
added initial prototype for deriving context free algebras from typeclasses

GrafBlutwurst committed:
fixed weird compiler warning

GrafBlutwurst committed:
fixed weird compiler warning

GrafBlutwurst committed:
changed to different style of pattern matching to work around weird type behaviour

GrafBlutwurst committed:
changed to different style of pattern matching to work around weird type behaviour

GrafBlutwurst committed:
changed contravatiant product case to use divisible rather than applicative (though it doesn't matter in this particular case)

GrafBlutwurst committed:
changed contravatiant product case to use divisible rather than applicative (though it doesn't matter in this particular case)

GrafBlutwurst committed:
added generic algebras, switched GenModule to use that

GrafBlutwurst committed:
added generic algebras, switched GenModule to use that

GrafBlutwurst committed:
merging new module structure

GrafBlutwurst committed:
adjusted to new module structure. sadly only possible with @unchecked annotation

GrafBlutwurst committed:
adjusted to new module structure. sadly only possible with @unchecked annotation

GrafBlutwurst committed:
minor refactor

GrafBlutwurst committed:
minor refactor

vil1 commented:
Kickly looking from my phone. This looks good, but it needs at least a test for the contravariant case (what about an interpreter for `Show`?)

GrafBlutwurst commented:
Good Idea. I was looking for a contravariant functor but couldn't come up with an example. show is great.

GrafBlutwurst committed:
impl of contravariant functor example for generic algebras. changes to API due to discoveries in said impl and some renamings

GrafBlutwurst committed:
impl of contravariant functor example for generic algebras. changes to API due to discoveries in said impl and some renamings

GrafBlutwurst committed:
added test

GrafBlutwurst committed:
added test

GrafBlutwurst commented:
the encoding of `Option[A]` as `A \/ Unit` does not yield very pretty results.

GrafBlutwurst committed:
fixed build.sbt formatting

GrafBlutwurst committed:
fixed build.sbt formatting

GrafBlutwurst committed:
adjusted json tests to conform with new encoder spec

GrafBlutwurst committed:
adjusted json tests to conform with new encoder spec

vil1 commented:
> the encoding of `Option[A]` as `A \/ Unit` does not yield very pretty results.

Mind sharing an example?

vil1 commented on build.sbt:
```diff
@@ -33,7 +45,7 @@ lazy val scalacheck = project
     )
   )
-  .dependsOn(core, `test-commons` % "test->test")
+  .dependsOn(core, generic, `test-commons` % "test->test")
```
I'd rather not make `scalacheck` (or any other module) depend on `generic` if possible (it feels like it breaks modularity).

That of course would prevent us to provide a `GenModule` based on `generic`, but I don't think it's very problematic (and a generic version of `GenModule` can still be implemented in `generic`'s tests).

That point remains open to discussion though.

vil1 commented on modules/generic/src/main/scala/GenericAlgebra.scala:
```diff
@@ -0,0 +1,76 @@
+  def covariantTargetFunctor[H[_]](
+    primNT: R.Prim ~> H,
+    seqNT: H ~> λ[X => H[List[X]]],
+    prodLabelNT: λ[X => (R.ProductTermId, H[X])] ~> H,
```
I suppose that if we rearrange the type parameters of `ProductTerm` so that `A` comes last, this NT could be be expressed (arguably) more simply as `ProductTerm[R.Prim, R.SumTermId, R.ProductTermId, H, ?] ~> H`.

Moreover, we could define useful type aliases in `SchemaModule` in the form of:
```scala
type RProductTerm[F[_], A] = ProductTerm[R.Prim, R.SumTermId, R.ProductTermId, F, A]
```

and further simplify our type signatures.

vil1 replied:
Also, maybe we should bundle these parameters into a single datatype.

GrafBlutwurst replied:
introduced `RProductTerm` and `RSumTerm` on `SchemaModule` and simplified the NTs accordingly. worked out nicely

vil1 commented on modules/generic/src/main/scala/ShowModule.scala:
```diff
@@ -0,0 +1,54 @@
+        val lhs = fa.shows(tpl._1)
+        val rhs = fb.shows(tpl._2)
+
+        if (lhs != "" && rhs != "") s"""($lhs, $rhs)"""
```
Cosmetics: I usually prefer to `match` on `(lhs, rhs)` in this kind of situation.

vil1 replied:
More nitpicking: I suppose it would be more efficient to use `show` and the `Cord#isEmpty` methods here.

GrafBlutwurst replied:
switched to using `Cord` and `isEmpty`. The matching however then requires guards unless there's a pattern for an empty `Cord` ?

vil1 reviewed (changes requested):
This looks pretty good overall!

My main concern is about modularity and the introduced dependency from `scalacheck` on `generic`, other comments are mostly cosmetics/nitpicking.

vil1 committed:
Rebased on prototyping and moved tests according to #39

GrafBlutwurst commented:
Frankly I'm also not quite happy with having to interject these generic derivations through inheritance. But I haven't found a good way to pass in the module in another way

GrafBlutwurst committed:
merging valentins PR

GrafBlutwurst committed:
Adressing PR Feedback
 - Simplyfying NT signatures for generic derivation
 - removing scalacheck module dependency on generic. instead adding a dedicated testcase
 - refactoring to use cord and patternmatch
 - removing some superflous files from test remodeling
 - cleaning up build sbt

GrafBlutwurst committed:
sbt formatting
